packages/langium/src/validation/validation-registry.ts

```diff
@@ -181,4 +205,45 @@ export class ValidationRegistry {
         return checks.map(entry => entry.check);
     }
 
+    /**
+     * Register logic which will be executed once before validating all the nodes of an AST/Langium document.
+     * This helps to prepare or initialize some information which are required or reusable for the following checks on the AstNodes.
+     *
+     * As an example, for validating unique fully-qualified names of nodes in the AST,
+     * here the map for mapping names to nodes could be established.
+     * During the usual checks on the nodes, they are put into this map with their name.
+     * @param checkBefore a set-up function which will be called once before actually validating an AST
+     * @param thisObj Optional object to be used as `this` when calling the validation check functions.
+     */
+    registerBeforeDocument(checkBefore: ValidationPreparation, thisObj: ThisParameterType<unknown> = this): void {
+        this.entriesBefore.push(this.wrapPreparationException(checkBefore, 'An error occurred during set-up of the validation', thisObj));
+    }
+
+    /**
+     * Register logic which will be executed once after validating all the nodes of an AST/Langium document.
+     * This helps to finally evaluate information which are collected during the checks on the AstNodes.
+     *
+     * As an example, for validating unique fully-qualified names of nodes in the AST,
+     * here the map with all the collected nodes and their names is checked
+     * and validation hints are created for all nodes with the same name.
+     * @param checkBefore a set-up function which will be called once before actually validating an AST
+     * @param thisObj Optional object to be used as `this` when calling the validation check functions.
+     */
+    registerAfterDocument(checkAfter: ValidationPreparation, thisObj: ThisParameterType<unknown> = this): void {
+        this.entriesAfter.push(this.wrapPreparationException(checkAfter, 'An error occurred during tear-down of the validation', thisObj));
+    }
+
+    protected wrapPreparationException(check: ValidationPreparation, messageContext: string, thisObj: unknown): ValidationPreparation {
+        return async (rootNode, accept, categories, cancelToken) => {
+            await this.handleException(() => check.call(thisObj, rootNode, accept, categories, cancelToken), messageContext, accept, rootNode);
+        };
+    }
+
+    getChecksBefore(): ValidationPreparation[] {
+        return this.entriesBefore;
+    }
+
+    getChecksAfter(): ValidationPreparation[] {
+        return this.entriesAfter;
+    }
```

Comment:
I'd declare the properties as `public readonly` rather than using such getter methods.

Reply:
The problem with `public` properties is, that users could register checks directly in this property and without calling `registerBeforeDocument(...)`, which bypasses our internal handling of exceptions.
Therefore I decided to keep the "simple getters".

Reply:
The getter does not change that: someone could still push elements to the returned array. If you really want to prevent that, you'd need to clone the array. But an easier and faster solution is just to document the property/method saying you shouldn't manipulate the array.

If you want to keep a getter, I suggest using a `get` property:
```typescript
    get checksAfter(): ValidationPreparation[] {
        return this.entriesAfter;
    }
```